### PR TITLE
[BROWSEUI] Add locked toolbar state persistence

### PR DESCRIPTION
**Pull request by cbialorucki**
## Purpose
Store the state of the explorer toolbar, locked or unlocked, in the proper registry location and set the proper toolbar state when initialized.

JIRA issue: [CORE-9094](https://jira.reactos.org/browse/CORE-9094)

## Proposed changes
- Set ```HKEY_CURRENT_USER\Software\Microsoft\Internet Explorer\Toolbar\Locked``` to 1 when the explorer toolbar is locked, 0 when the toolbar is unlocked.
- Query ```HKEY_CURRENT_USER\Software\Microsoft\Internet Explorer\Toolbar\Locked``` when the toolbar is initialized to set the correct locked state.
- Set the default state of the toolbar to locked if the registry key does not exist, matching the behavior of Windows Server 2003.

### CODE REVIEW

File: dll/win32/browseui/internettoolbar.cpp [hunk 1]
```diff
@@ -715,9 +731,21 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
     int                                     bandCount;
     CDockSite                               *dockSite;
     HRESULT                                 hResult;
+    HKEY                                    hKey;
 
     if (locked != fLocked)
     {
+        if(RegCreateKeyEx(HKEY_CURRENT_USER,
```
Comment (HBelusca) [inline dll/win32/browseui/internettoolbar.cpp:738]: Please change (in this PR):
1. Space between `if` and opening parenthesis;
2. Use explicit unicode functions (-W and `L"strings"`);
3. indentation.

File: dll/win32/browseui/internettoolbar.cpp [hunk 2]
```diff
@@ -607,8 +607,24 @@ HRESULT STDMETHODCALLTYPE CMenuCallback::CallbackSM(LPSMDATA psmd, UINT uMsg, WP
...
+    if(RegOpenKeyEx(HKEY_CURRENT_USER,
+                    _T("Software\\Microsoft\\Internet Explorer\\Toolbar"),
+                    0, KEY_READ, &hKey) == ERROR_SUCCESS)
+    {
+        DWORD dwLocked;
+        DWORD dwSize;
+        if(RegQueryValueEx(hKey,
+                        _T("Locked"),
+                        NULL, NULL, (LPBYTE)&dwLocked, &dwSize) == ERROR_SUCCESS)
+        {
```
Comment (HBelusca) [inline dll/win32/browseui/internettoolbar.cpp:622]: Maybe check also that the retrieved value is of type `REG_DWORD` and the `dwSize == sizeof(dwLocked)` .

Comment (cbialorucki) [reply dll/win32/browseui/internettoolbar.cpp:628]: Thank you for the suggestion! I added checks to ensure the retrieved registry key is the correct size and type.

File: dll/win32/browseui/internettoolbar.cpp (hunk 1 above)
Comment (cbialorucki) [reply dll/win32/browseui/internettoolbar.cpp]: Thank you for reviewing my pull request so quickly! I have added a new commit implementing your suggestions in accordance with our [coding style](https://reactos.org/wiki/Coding_Style).  If you want me to make any more changes, please let me know!

File: dll/win32/browseui/internettoolbar.cpp [hunk 3]
```diff
@@ -715,9 +740,29 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
     int                                     bandCount;
     CDockSite                               *dockSite;
     HRESULT                                 hResult;
+    HKEY                                    hKey;
 
     if (locked != fLocked)
     {
+        if (RegCreateKeyExW(HKEY_CURRENT_USER,
```
Comment (JoachimHenze) [inline dll/win32/browseui/internettoolbar.cpp:747]:
```suggestion
        if (RegCreateKeyExW(HKEY_CURRENT_USER,
```

File: dll/win32/browseui/internettoolbar.cpp [hunk 4]
```diff
@@ -715,9 +740,29 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
     int                                     bandCount;
     CDockSite                               *dockSite;
     HRESULT                                 hResult;
+    HKEY                                    hKey;
 
     if (locked != fLocked)
     {
+        if (RegCreateKeyExW(HKEY_CURRENT_USER, 
+                            L"Software\\Microsoft\\Internet Explorer\\Toolbar",
```
Comment (JoachimHenze) [inline dll/win32/browseui/internettoolbar.cpp:748]:
```suggestion
                            L"Software\\Microsoft\\Internet Explorer\\Toolbar",
```

File: dll/win32/browseui/internettoolbar.cpp [hunk 5]
```diff
@@ -715,9 +740,29 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
...
+                            0,
+                            NULL,
+                            REG_OPTION_NON_VOLATILE, 
+                            KEY_WRITE,
+                            NULL,
+                            &hKey,
+                            NULL) == ERROR_SUCCESS)
+        {
+            DWORD dwLocked = (locked ? 1 : 0);
+            RegSetValueExW(hKey,
```
Comment (JoachimHenze) [inline dll/win32/browseui/internettoolbar.cpp:758]:
```suggestion
            RegSetValueExW(hKey,
```

**Review by JoachimHenze — CHANGES_REQUESTED** [review]
Some EOL whitespace should be stripped.

**Comment by cbialorucki** [comment]
> Some EOL whitespace should be stripped.

EOL whitespace removed.

File: dll/win32/browseui/internettoolbar.cpp [hunk 6]
```diff
@@ -607,8 +607,33 @@ HRESULT STDMETHODCALLTYPE CMenuCallback::CallbackSM(LPSMDATA psmd, UINT uMsg, WP
 
 CInternetToolbar::CInternetToolbar()
 {
+    HKEY hKey;
+    // Sets the toolbar to locked by default if the registry key is missing.
```
Comment (HBelusca) [inline dll/win32/browseui/internettoolbar.cpp:611]:
```suggestion

    // Set the toolbar to locked by default if the registry key is missing.
```

File: dll/win32/browseui/internettoolbar.cpp [hunk 7]
```diff
@@ -607,8 +607,33 @@ HRESULT STDMETHODCALLTYPE CMenuCallback::CallbackSM(LPSMDATA psmd, UINT uMsg, WP
...
+                             (LPBYTE)&dwLocked,
+                             &dwSize) == ERROR_SUCCESS)
+        {
+            if (dwType == REG_DWORD && dwSize == sizeof(dwLocked))
+            {
+                fLocked = (dwLocked != 0);
+            }
+        }
+    }
+    RegCloseKey(hKey);
```
Comment (HBelusca) [inline dll/win32/browseui/internettoolbar.cpp:635]:
```suggestion
    RegCloseKey(hKey);

```

File: dll/win32/browseui/internettoolbar.cpp [hunk 8]
```diff
@@ -715,9 +740,29 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
...
+            DWORD dwLocked = (locked ? 1 : 0);
+            RegSetValueExW(hKey,
+                           L"Locked", 
+                           0,
+                           REG_DWORD,
+                           (const BYTE*)&dwLocked,
+                           sizeof(dwLocked));
+        }
+        RegCloseKey(hKey);
         fLocked = locked;
```
Comment (HBelusca) [inline dll/win32/browseui/internettoolbar.cpp:766]:
```suggestion
        fLocked = locked;

```

File: dll/win32/browseui/internettoolbar.cpp [hunk 9]
```diff
@@ -715,9 +740,29 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
...
+                            NULL,
+                            &hKey,
+                            NULL) == ERROR_SUCCESS)
+        {
+            DWORD dwLocked = (locked ? 1 : 0);
+            RegSetValueExW(hKey,
+                           L"Locked", 
+                           0,
+                           REG_DWORD,
+                           (const BYTE*)&dwLocked,
```
Comment (HBelusca) [inline dll/win32/browseui/internettoolbar.cpp:762]: Question: Is the cast here necessary?

File: dll/win32/browseui/internettoolbar.cpp [hunk 10]
```diff
@@ -607,8 +607,33 @@ HRESULT STDMETHODCALLTYPE CMenuCallback::CallbackSM(LPSMDATA psmd, UINT uMsg, WP
...
+                      &hKey) == ERROR_SUCCESS)
+    {
+        DWORD dwLocked;
+        DWORD dwSize;
+        DWORD dwType;
+        if (RegQueryValueExW(hKey,
+                             L"Locked",
+                             NULL,
+                             &dwType,
+                             (LPBYTE)&dwLocked,
```
Comment (HBelusca) [inline dll/win32/browseui/internettoolbar.cpp:626]: Question: Is the cast here necessary?

**Comment by HBelusca** [comment]
lgtm otherwise.

File: dll/win32/browseui/internettoolbar.cpp [hunk 11]
```diff
@@ -715,10 +742,31 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
...
     HRESULT                                 hResult;
+    HKEY                                    hKey;
 
     if (locked != fLocked)
     {
+        if (RegCreateKeyExW(HKEY_CURRENT_USER,
+                            L"Software\\Microsoft\\Internet Explorer\\Toolbar",
+                            0,
+                            NULL,
+                            REG_OPTION_NON_VOLATILE,
```
Comment (JoachimHenze) [inline dll/win32/browseui/internettoolbar.cpp:753]:
```suggestion
                            REG_OPTION_NON_VOLATILE,
```

Comment (JoachimHenze) [inline dll/win32/browseui/internettoolbar.cpp:753]:
```suggestion
                            REG_OPTION_NON_VOLATILE,
```

File: dll/win32/browseui/internettoolbar.cpp [hunk 12]
```diff
@@ -715,10 +742,31 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
...
+                            NULL,
+                            REG_OPTION_NON_VOLATILE, 
+                            KEY_WRITE,
+                            NULL,
+                            &hKey,
+                            NULL) == ERROR_SUCCESS)
+        {
+            DWORD dwLocked = (locked ? 1 : 0);
+            RegSetValueExW(hKey,
+                           L"Locked",
```
Comment (JoachimHenze) [inline dll/win32/browseui/internettoolbar.cpp:761]:
```suggestion
                           L"Locked",
```

**Review by JoachimHenze — CHANGES_REQUESTED** [review]

File: dll/win32/browseui/internettoolbar.cpp (hunk 11 above)
Comment (cbialorucki) [reply dll/win32/browseui/internettoolbar.cpp]: EOL whitespace removed in last commit.

**Review by JoachimHenze — APPROVED** [review]
Approved. Please try to answer the questions of hbelusca regarding the casts before merging.

File: dll/win32/browseui/internettoolbar.cpp [hunk 13]
```diff
@@ -607,8 +607,35 @@ HRESULT STDMETHODCALLTYPE CMenuCallback::CallbackSM(LPSMDATA psmd, UINT uMsg, WP
 
 CInternetToolbar::CInternetToolbar()
 {
+    HKEY hKey;
+
+    // Set the toolbar to locked by default if the registry key is missing.
+    fLocked = true;
+    if (RegOpenKeyExW(HKEY_CURRENT_USER,
```
Comment (learn-more) [inline dll/win32/browseui/internettoolbar.cpp:614]: Further down the file the function `SHRegGetBoolUSValueW` is used, any particular reason why you did not use that here?

File: dll/win32/browseui/internettoolbar.cpp [hunk 14]
```diff
@@ -715,10 +719,31 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
     int                                     bandCount;
     CDockSite                               *dockSite;
     HRESULT                                 hResult;
+    HKEY                                    hKey;
 
     if (locked != fLocked)
     {
+        if (RegCreateKeyExW(HKEY_CURRENT_USER,
```
Comment (learn-more) [inline dll/win32/browseui/internettoolbar.cpp:726]: I was hoping commenting on one function would be enough, but any reason why you are not using `SHRegSetUSValueW` here?

File: dll/win32/browseui/internettoolbar.cpp [hunk 15]
```diff
@@ -718,7 +722,15 @@ HRESULT CInternetToolbar::LockUnlockToolbars(bool locked)
 
     if (locked != fLocked)
     {
+        DWORD dwLocked = locked ? 1 : 0;
+        SHRegSetUSValueW(L"Software\\Microsoft\\Internet Explorer\\Toolbar",
+                         L"Locked",
+                         REG_DWORD,
+                         &dwLocked,
+                         sizeof(DWORD),
```
Comment (binarymaster) [inline dll/win32/browseui/internettoolbar.cpp:730]: Wouldn't it just work with BOOL, i.e. without additional DWORD variable?

```suggestion
        SHRegSetUSValueW(L"Software\\Microsoft\\Internet Explorer\\Toolbar",
                         L"Locked",
                         REG_DWORD,
                         &locked,
                         sizeof(locked),
```

Comment (yagoulas) [reply dll/win32/browseui/internettoolbar.cpp:730]: In my opinion it is better with it being DWORD.

**Review by yagoulas — APPROVED** [review]

File: dll/win32/browseui/internettoolbar.cpp (hunk 15 above)
Comment (HBelusca) [inline dll/win32/browseui/internettoolbar.cpp:730]:
```suggestion
                         sizeof(dwLocked),
```

**Review by HBelusca — APPROVED** [review]

**Review by learn-more — APPROVED** [review]

**Comment by HBelusca** [comment]
Thanks for the PR!

**Pull request merged**